pgxn/neon/walproposer.c

```
@@ -103,72 +104,122 @@ WalProposerCreate(WalProposerConfig *config, walproposer_api api)
 
-	for (; host != NULL && *host != '\0'; host = sep)
+	for (; connstring != NULL && *connstring != '\0'; connstring = sep)
 	{
-		port = strchr(host, ':');
-		if (port == NULL)
-		{
-			wp_log(FATAL, "port is not specified");
-		}
-		*port++ = '\0';
-		sep = strchr(port, ',');
+		char	   *port;
+		char	   *errmsg;
+		Safekeeper *sk;
+		PQconninfoOption *conninfo_options, *option;
+
+		sk = &wp->safekeeper[wp->n_safekeepers];
+
+		sep = strchr(connstring, ',');
 		if (sep != NULL)
 			*sep++ = '\0';
 		if (wp->n_safekeepers + 1 >= MAX_SAFEKEEPERS)
 		{
 			wp_log(FATAL, "too many safekeepers");
 		}
-		wp->safekeeper[wp->n_safekeepers].host = host;
-		wp->safekeeper[wp->n_safekeepers].port = port;
-		wp->safekeeper[wp->n_safekeepers].state = SS_OFFLINE;
-		wp->safekeeper[wp->n_safekeepers].active_state = SS_ACTIVE_SEND;
-		wp->safekeeper[wp->n_safekeepers].wp = wp;
+
+		/* TODO(tristan957): Remove this compatibility code and only keep the
+		 * else branch.
+		 *
+		 * Check if we have a connection string formatted as host:port, and use
+		 * the old parsing code.
+		 */
+		port = strchr(connstring, ':');
+		if (port)
+		{
+			sk->host = connstring;
+			*port++ = '\0';
+			sk->port = port;
+		}
+		else
+		{
+			conninfo_options = PQconninfoParse(connstring, &errmsg);
+			if (!conninfo_options)
+				wp_log(FATAL, "invalid safekeeper connection string: %s", errmsg);
+
+			// Save off the host and port for identification purposes
+			option = conninfo_options;
+			while (option)
+			{
+				if (!option->keyword)
+					break;
+
+				if (strcmp(option->keyword, "host") == 0)
+				{
+					sk->host = pstrdup(option->val);
+					goto end;
+				}
+
+				if (strcmp(option->keyword, "port") == 0)
+				{
+					sk->port = pstrdup(option->val);
+					goto end;
+				}
+
+			end:
+				// We've saved both the host and port, so we can skip iterating
+				// the rest of the list
+				if (sk->host && sk->port)
+					break;
+
+				option++;
+			}
```

What if we didn't find `host` or `port` at all? Shouldn't we return an error somehow?
I guess we will dereference NULL somewhere later if we don't